source/length.js

```diff
@@ -3,21 +3,33 @@ import _isNumber from './internal/_isNumber';
 
 
 /**
- * Returns the number of elements in the array by returning `list.length`.
+ * Returns the number of elements in the array, set or map
+ * by returning `collection.length` or `collection.size`.
  *
  * @func
  * @memberOf R
  * @since v0.3.0
  * @category List
  * @sig [a] -> Number
- * @param {Array} list The array to inspect.
- * @return {Number} The length of the array.
+ * @param {Array|Set|Map} collection The collection to inspect.
+ * @return {Number} The length of the collection.
  * @example
  *
  *      R.length([]); //=> 0
  *      R.length([1, 2, 3]); //=> 3
+ *      R.length(new Set(['x', 'y', 'z'])); //=> 3
+ *      R.length(new Map([['k1', 'v1'], ['k2', 'v2']])); //=> 2
  */
 var length = _curry1(function length(list) {
-  return list != null && _isNumber(list.length) ? list.length : NaN;
+  if (list == null) {
+    return NaN;
+  }
+  if (_isNumber(list.length)) {
+    return list.length;
+  }
+  if (_isNumber(list.size)) {
+    return list.size;
+  }
+  return NaN;
```

Comment:
Should we also handle the `Object` type?  How much difference is there between saying that a `Map` has a size and saying that an `Object` has one?

Reply:
I know that this can be done well, but I'm not sure if it should be done.

Reading a property is `O(1)`, iterating over object is `O(n)`.

In the interest of not getting this PR stuck in a limbo, I suggest to not do it in this PR.
If someone thinks this is a good idea, they can open a new issue or PR just for that.

Reply:
I don't know how the engines implement this, but there is a non-trivial chance that `Object.keys` does not involve iterating anything. 

>  In the interest of not getting this PR stuck in a limbo,...

Unfortunately, it seems a great number of issues and PRs are trapped there.  But if we're going to change this function, rather than just remove it, I'd prefer to do it once.